cmd/admin-handlers.go

```diff
@@ -207,3 +211,72 @@ func (adminAPI adminAPIHandlers) ClearLocksHandler(w http.ResponseWriter, r *htt
 	// Reply with list of locks cleared, as json.
 	writeSuccessResponse(w, jsonBytes)
 }
+
+func validateHealQueryParams(vars url.Values) (string, string, string, string, int, APIErrorCode) {
+	bucket := vars.Get(string(mgmtBucket))
+	prefix := vars.Get(string(mgmtPrefix))
+	marker := vars.Get(string(mgmtMarker))
+	delimiter := vars.Get(string(mgmtDelimiter))
+	maxKeyStr := vars.Get(string(mgmtMaxKey))
+
+	// N B empty bucket name is invalid
+	if !IsValidBucketName(bucket) {
+		return "", "", "", "", 0, ErrInvalidBucketName
+	}
+
+	// empty prefix is valid.
+	if !IsValidObjectPrefix(prefix) {
+		return "", "", "", "", 0, ErrInvalidObjectName
+	}
+
+	// check if maxKey is a valid integer.
+	maxKey, err := strconv.Atoi(maxKeyStr)
+	if err != nil {
+		return "", "", "", "", 0, ErrInvalidMaxKeys
+	}
+
+	// Validate prefix, marker, delimiter and maxKey.
+	apiErr := validateListObjectsArgs(prefix, marker, delimiter, maxKey)
+	if apiErr != ErrNone {
+		return "", "", "", "", 0, apiErr
+	}
+
+	return bucket, prefix, marker, delimiter, maxKey, ErrNone
+}
+
+// ListObjectsHealHandler - GET /?heal&bucket=mybucket&prefix=myprefix&marker=mymarker&delimiter=&mydelimiter&maxKey=1000
+// - bucket is mandatory query parameter
+// - rest are optional query parameters
+// List upto maxKey objects that need healing in a given bucket matching the given prefix.
+func (adminAPI adminAPIHandlers) ListObjectsHealHandler(w http.ResponseWriter, r *http.Request) {
+	adminAPIErr := checkRequestAuthType(r, "", "", "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, r, adminAPIErr, r.URL.Path)
+		return
+	}
+
+	vars := r.URL.Query()
+	bucket, prefix, marker, delimiter, maxKey, adminAPIErr := validateHealQueryParams(vars)
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, r, adminAPIErr, r.URL.Path)
+		return
+	}
+
+	objLayer := newObjectLayerFn()
+	if objLayer == nil {
```

Comment:
Return this error in the beginning, don't validate incoming arguments if you can't serve them.